### PR TITLE
Corrected default values of xextent in specgram(). Fixes Bug #7666.

### PR DESCRIPTION
**Pull request by vollbier**
Let's the time axis start from the midpoint of the first segment instead at 0. Details are in the bug report #7666.

Todo:
- [x] Deal with failing image comparison tests.
- [x] Pad the sides

### CODE REVIEW

**Comment by tacaswell** [comment]
Should these be padded out to be the left and right edges of the first and last segments so that the mid points of the pixels are the midpoint of the time bins?

**Comment by vollbier** [comment]
@tacaswell, I'm not quite sure, if I understand what you mean by padding: The length in of the spectrogram is always shorter then the signal (by `nfft-1` samples in my example). With the fix, the point in time is exactly at the midpoint of the segment (i.e., window function).

**Comment by tacaswell** [comment]
The `extent` kwarg sets where the edges of the image will be placed.  If `t` is the midpoints, then this starts the first pixel at the mid point of the first time window and the last pixel ends at the midpoint of the last bin.  I think the correct behavior should be that each pixel starts and stops on the edges of the time window represents (or is at least centered on the time window if there is overlap).

**Comment by vollbier** [comment]
Ah, now I understand:
specgram_nov_correct
specgram_nov_default
Never noticed that, because I usually try to maximize the overlap.

**Comment by vollbier** [comment]
Padding Fixed:
specgram_nov0
specgram_nov255

**Comment by tacaswell** [comment]
Thanks, LGTM to me!

This is another case where we would have asked you to squash and re-base (to only have one set of the regenerated images in the repository history), but in general you do not have to squash down to a single commit.

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -7228,7 +7229,9 @@ def specgram(self, x, NFFT=None, Fs=None, Fc=None, detrend=None,
         Z = np.flipud(Z)
 
         if xextent is None:
-            xextent = 0, np.max(t)
+            # padding is needed for first and last segment:
+            pad_xextnt = (NFFT-noverlap) / Fs / 2
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: `pad_xextent`.

Comment (vollbier) [inline lib/matplotlib/axes/_axes.py]: Done.

**Comment by tacaswell** [comment]
Thanks @vollbier 🎉 

I think this is your first contribution to mpl, hopefully there will be many more!

**Comment by vollbier** [comment]
I hope so too.

**Comment by QuLogic** [comment]
> @tacaswell added this to the 2.1 (next point release) milestone 4 days ago

#7666 says 2.0.1?

**Comment by tacaswell** [comment]
Bumped the issue to 2.1, I think this change is too big to put into 2.0 (at least on top of the rc).  I am very open to be convinced otherwise.

**Comment by vollbier** [comment]
@tacaswell I have no idea, how such changes disturb matplotlib's release cycles. From the user's point of view, I would tend to include it: I stumbled upon this issue while comparing spectrograms to Wigner-Ville distributions for certain signals. It lead me at first to the wrong conclusion, that spectrograms give biased estimates, which can be a quite serious roadblock for further investigations.

**Comment by tacaswell** [comment]
@vollbier sold.  

backported as v2.x 6afa9cef1

**Pull request merged**